streaming/base/util.py

```diff
@@ -209,6 +215,120 @@ def get_import_exception_message(package_name: str, extra_deps: str) -> str:
             f'`pip install \'mosaicml-streaming[{package_name}]\'`.'
 
 
+def do_merge_index(index_file_urls: List[Any],
+                   out: Union[str, Tuple[str, str]],
+                   keep_local: bool = True,
+                   download_timeout: int = 60) -> None:
+    """Merge index.json from a list of index.json. Write to `out`, overwriting if exists.
+
+    Args:
+        index_file_urls (Union[str, Tuple[str,str]]): index.json from all the partitions
+            each element can take the form of a single path string or a tuple string.
+
+            The pattern of index_file_urls and corresponding reaction is one of:
+            1. All urls are str (local). All urls are accessible locally -> no download
+            2. All urls are tuple (local, remote). All urls are accessible locally -> no download
+            3. All urls are tuple (local, remote). At least one url is not accessible locally -> download all
```

Comment:
could you clarify why we need to download all index files if just one url is not accessible locally?

Reply:
Yeah good point. Can we just download if it is not present locally?

Reply:
This is to keep things simple so I don't mess up the folder structure local-remote. so it's either "no download" or "download them all".

Reply:
rewrite to allow "only download missing ones to local".

Reply:
thank you!!

Reply:
Good catch. Fixed.